src/utils/logger.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Abstract overridable class for logger
+ */
+export abstract class ILogger {
+  constructor(protected readonly _logLevel: LogLevel = LogLevel.debug) {}
+
+  abstract log(message: unknown): void;
+  abstract warn(message: unknown): void;
+  abstract error(message: unknown): void;
+  abstract debug(message: unknown): void;
+}
+
+export enum LogLevel {
+  debug = 1,
+  log = 2,
+  warn = 3,
+  error = 4,
+}
+
+/*
+ * Default Implementation of ILogger streaming logs to console
+ */
+export class ConsoleLogger extends ILogger {
```

Comment:
A small optimization: we can filter logs in base class and in subclasses implement only specific logger feature. For example `ILogger.warn` could be
`if (this._logLevel <=LogLevel.warn) {
   this.log(message)
`
So the only abstract method is `log`. Also method currently name `log` can be renamed `info`

Reply:
this would require to duplicate all functions but I'm up for that. We would have 
- 4 non abstract functions : `info(), warn(), debug(), error()` containing log level filter and 
- 4 abstract private functions `logInfo(), logWarn(), logDebug(), logError()` containing only logging functionality.   

I this how you imagin, too ?

Reply:
Not actually. I mean there is only one abstract method - the one which is actually does logging. And only this method is overridden in sub-classes. Like this:
```
abstract class ILogger {
   abstract _log(message); // only this is overridden

   info(message) {
      if (this._logLevel <= LogLevel.Info) {
         this.**_log**(message);
      }
   }

   warn(message) {
      if (this._logLevel <= LogLevel.Warn) {
         this.**_log**(message);
      }
   }
  ...
}

class ConsolLoger extends ILogger {
   _log(message) {
       console.log(message);
   }
}

```

Reply:
but how would we know if this is error or warn or anything else if the only thing we always log is  log(message). They usually differ with format , eg `{LogLevel} - {date} - {message}` , sometimes `color` how would you achieve it with your solution?